src/libraries/Common/src/System/Reflection/Metadata/TypeNameParser.cs

```diff
@@ -0,0 +1,257 @@
+                }
+
+                // For [[, there had better be a ']' after the type name.
+                if (doubleBrackets && !TryStripFirstCharAndTrailingSpaces(ref _inputString, ']'))
+                {
+                    return null;
+                }
+
+                if (genericArgs is null)
+                {
+#if SYSTEM_PRIVATE_CORELIB
+                    genericArgs = new List<TypeName>(2);
+#else
+                    genericArgs = ImmutableArray.CreateBuilder<TypeName>(2);
+#endif
+                }
+                genericArgs.Add(genericArg);
+
+                // Is there a ',[' indicating fully qualified generic type arg?
+                // Is there a ',' indicating non-fully qualified generic type arg?
+                if (TryStripFirstCharAndTrailingSpaces(ref _inputString, ','))
+                {
+                    doubleBrackets = TryStripFirstCharAndTrailingSpaces(ref _inputString, '[');
+
+                    goto ParseAnotherGenericArg;
+                }
+
+                // The only other allowable character is ']', indicating the end of
+                // the generic type arg list.
+                if (!TryStripFirstCharAndTrailingSpaces(ref _inputString, ']'))
+                {
+                    return null;
+                }
+            }
+
+            // If there was an error stripping the generic args, back up to
+            // before we started processing them, and let the decorator
+            // parser try handling it.
+            if (genericArgs is null)
+            {
+                _inputString = capturedBeforeProcessing;
+            }
+
+            int previousDecorator = default;
+            // capture the current state so we can reprocess it again once we know the AssemblyName
+            capturedBeforeProcessing = _inputString;
+            // iterate over the decorators to ensure there are no illegal combinations
+            while (TryParseNextDecorator(ref _inputString, out int parsedDecorator))
+            {
+                if (!TryDive(ref recursiveDepth))
+                {
+                    return null;
+                }
+
+                // Currently it's illegal for managed reference to be followed by any other decorator,
+                // but this is a runtime-specific behavior and the parser is not enforcing that rule.
+                previousDecorator = parsedDecorator;
+            }
+
+            AssemblyNameInfo? assemblyName = null;
+            if (allowFullyQualifiedName && !TryParseAssemblyName(ref assemblyName))
+            {
+#if SYSTEM_PRIVATE_CORELIB
+                // backward compat: throw FileLoadException for non-empty invalid strings
+                if (!_throwOnError && _inputString.TrimStart().StartsWith(","))
+                {
+                    return null;
+                }
+                throw new IO.FileLoadException(SR.InvalidAssemblyName, _inputString.ToString());
+#else
+                return null;
+#endif
```

Comment:
```suggestion
#if SYSTEM_PRIVATE_CORELIB
                // backward compat: throw FileLoadException for non-empty invalid strings
                if (_throwOnError || !_inputString.TrimStart().StartsWith(","))
                {
                    throw new IO.FileLoadException(SR.InvalidAssemblyName, _inputString.ToString());
                }
#endif

                return null;
```
?

Reply:
Also, we don't want to throw on error if _throwOnError is true, even for non-corelib?

Reply:
We are going to throw, but `ArgumentException` rather than `FileLoadException` (to be consistent, invalid input == argument exception)